lib/logflare/backends/adaptor/clickhouse_adaptor.ex

```diff
@@ -134,6 +84,47 @@ defmodule Logflare.Backends.Adaptor.ClickhouseAdaptor do
     |> validate_user_pass()
   end
 
+  # delegated functions
+  defdelegate ch_connection_count(), to: Supervisor
+  defdelegate ch_connection_exists?(backend), to: Supervisor
+
+  defdelegate execute_ch_query(backend_or_conn, statement, params \\ [], opts \\ []),
+    to: Client
+
+  defdelegate find_or_create_ch_connection(backend), to: Supervisor
+  defdelegate insert_log_event(source, backend, log_event), to: Client
+  defdelegate insert_log_events(source, backend, log_events), to: Client
+  defdelegate terminate_ch_connection(backend), to: Supervisor
+
+  @impl GenServer
+  def init({source, backend}) do
+    # establish connection
+    case find_or_create_ch_connection(backend) do
+      {:ok, _pid} ->
+        :ok
+
+      res ->
+        Logger.error("Failed to create ClickHouse connection: #{inspect(res)} ",
+          source_token: source.token,
+          backend_id: backend.id
+        )
+    end
+
+    # maybe handle table creation or permission checks here?
+
+    state = %__MODULE__{
+      config: backend.config,
+      backend: backend,
+      backend_token: if(backend, do: backend.token, else: nil),
+      source_token: source.token,
+      source: source,
+      pipeline_name: Backends.via_source(source, Pipeline, backend.id)
+    }
+
+    {:ok, _pipeline_pid} = Pipeline.start_link(state)
```

Comment:
instead of having a separate DynamicSupervisor to manage all CH connections, would it be feasible to start the connection here together with Broadway here? 
Name registry can then use the usual `Backends.via_source/3` functions.
Although the PostgresAdaptor uses a separate DynamicSupervisor, but it would be much better to encapsulate all related procs under the same Adaptor process tree so that when we reset proc state, it would not leave stale procs hanging in the DynSup.

Reply:
this has revealed a architecture issue with the PostgresAdaptor :P

Reply:
@Ziinc I have updated this implementation by removing those additional layers. Still using a GenServer for the adaptor though.

Reply:
I've added some logic to leverage the `DBConnection` [`:after_connect`](https://hexdocs.pm/db_connection/DBConnection.html#start_link/2) option for provisioning the ClickHouse table.

This works fine, but I notice this logic fires as many times as pools that are configured. Will investigate how to prevent that or create a workaround when I have a chance.

Reply:
I have dropped the `after_connect` approach because it does fire per connection - so a pool size of 3 kicks off that logic 3 times.

Went with a temp GenServer approach instead 🤷